Feca 12 most selling products

YassenAhmed-gif opened the pull request.

YassenAhmed-gif committed:
feat: Implement best seller feature

This commit introduces the best seller feature, including API integration, data handling, and domain logic.

Key changes include:
- Added `BestSellerRemoteDataSourceImpl` to handle API requests for best seller products.
- Implemented `ApiServices` for defining best seller-related API endpoints.
- Updated `pubspec.lock` with new dependencies.
- Added `BestSellerRepoImpl` to manage best seller-related data operations.
- Defined `BestSellerRepo` interface for abstracting best seller repositories.
- Defined `BestSellerRemoteDataSource` interface for abstracting best seller data sources.
- Added `bestSeller` endpoint to `AppConstants`.
- Created `BestSellerResponseEntity` to represent the best seller response entity.
- Created `BestSellerEntity` to represent a best seller product.
- Implemented `BestSellerDtoMapper` and `BestSellerResponseDtoMapper` for mapping DTOs to entities.
- Created `BestSellerDto` and `BestSellerResponseDto` for structuring API data.
- Added `BestSellerUseCase` for handling the best seller business logic.
- Removed unused authentication-related files.
- Updated `pubspec.yaml` to include `bloc_test` and remove commented-out sections.

YassenAhmed-gif committed:
feat: Add BestSellerResponseEntity

This commit introduces the `BestSellerResponseEntity` class.

Key changes include:
- Created `BestSellerResponseEntity` to represent the best seller response entity.

YassenAhmed-gif committed:
Refactor: Implement Best Seller Feature

This commit introduces the Best Seller feature, allowing users to view a list of best-selling products.

Key changes include:

- **Feature Restructuring:**
    - Moved Best Seller related files from `lib/Feature/mainLayout/tabs/bestSellerFeature/` to `lib/Feature/bestSellerFeature/`.
    - Removed `lib/Feature/mainLayout/main_layout.dart`.

- **UI Implementation:**
    - Created `BestSellerScreen` to display best-selling products in a grid view.
    - Added `ProductCard` widget for a reusable product display component.
    - Created `ProductDetailsScreen` (placeholder).

- **State Management (BLoC):**
    - Implemented `BestSellerViewModel`, `BestSellerState`, and `BestSellerEvent` for managing the Best Seller screen's state and logic.

- **Data Layer:**
    - Simplified `BestSellerDto` and `BestSellerEntity` to include essential product information and added a `discountPercent` getter to `BestSellerEntity`.
    - Updated mappers (`BestSellerDtoMapper`, `BestSellerResponseDtoMapper`) accordingly.
    - Refined `BestSellerRepoImpl` to correctly return the result from the data source.

- **Dependency Injection & Services:**
    - Updated `DioModule` to use `Storage` interface with `AppConstants.secureStorage` instance name for token retrieval.
    - Corrected `Storage` interface method `read` to return `Future<String>`.
    - Renamed `AppConstants.secureStorge` to `AppConstants.secureStorage`.

- **Navigation:**
    - Added routes for `BestSellerScreen` and `ProductDetailsScreen` in `AppRoutes` and `RouteGenerator`.
    - Set `BestSellerScreen` as the initial route in `main.dart` for development/testing purposes.

- **Localization:**
    - Added English and Arabic translations for "Best Seller", "Bloom with our exquisite best sellers", "add to cart", and "EGP" in `en.json` and `ar.json`.

- **Core Widgets & Utilities:**
    - Added new size constants in `sizes.dart` for UI elements.
    - Corrected a color constant in `app_colors.dart`.

YassenAhmed-gif committed:
Refactor: Update localization in Best Seller screen

This commit updates the `BestSellerScreen` to use `easy_localization` for the title text.

Key changes:
- Imported `easy_localization` package.
- Updated `Text` widget to use `LocaleKeys.bestSeller.tr()` for localized title.

YassenAhmed-gif committed:
feat: Add shimmer effect for loading states

This commit introduces a shimmer loading animation to enhance the user experience during data fetching.

Key changes:
- Added `ListViewsShimmerWidget` to display a placeholder loading UI for lists.
- Integrated `ListViewsShimmerWidget` into `BestSellerScreen` to show a shimmer effect while best seller products are loading.
- Added `shimmer_animation` dependency to `pubspec.yaml` and `pubspec.lock`.

YassenAhmed-gif committed:
feat: Implement best seller feature

This commit introduces the best seller feature, including API integration and localization.

Key changes include:
- Added generated files for API service, best seller DTO, and best seller response DTO to `.gitignore`.
- Added generated files for dependency injection configuration and localization keys to `.gitignore`.

YassenAhmed-gif committed:
Refactor: Update .gitignore

This commit updates the .gitignore file to include generated files related to various features.

Key changes:
- Ignored `pubspec.lock`.
- Ignored generated files for `categoriesFeature` including API client, DTOs, and response models.
- Ignored generated files for `occasionFeature` including API client, DTOs, response models, and test mocks.
- Ignored generated files for `authFeature` including request and response DTOs.

YassenAhmed-gif committed:
Test: Add tests for BestSellerFeature and ProductCard widget

This commit introduces a comprehensive suite of tests for the BestSellerFeature and the ProductCard widget.

Key changes:
- Added `product_card_test.dart` to verify the ProductCard widget's UI and functionality, including correct display of product information and handling of long titles.
- Added `best_seller_feature_test.dart` with unit tests for:
    - `BestSellerEntity`: Verifying discount calculation.
    - `BestSellerRemoteDataSourceImpl`: Testing success and error scenarios for API calls.
    - `BestSellerRepoImpl`: Ensuring correct data propagation from the data source.
    - `BestSellerUseCase`: Testing the use case's interaction with the repository.
    - `BestSellerViewModel`: Using `bloc_test` to verify state changes based on events (success and failure cases for fetching best sellers).
    - `BestSellerDto` to `BestSellerEntity` mapping.
    - `BestSellerState`: Testing the `copyWith` method.
- Updated `.gitignore` to include the generated mock file for `best_seller_feature_test.dart`.
- Renamed `besst_seller_repo_impl.dart` to `best_seller_repo_impl.dart` to fix a typo.

ali-safwat commented on lib/Feature/bestSellerFeature/api/dataSources/best_seller_remote_data_source_impl.dart:
```diff
@@ -0,0 +1,29 @@
+
+@Injectable(as: BestSellerRemoteDataSource)
+class BestSellerRemoteDataSourceImpl implements BestSellerRemoteDataSource {
+  final ApiServices apiServices;
```
Make apiServices private inside this class to enforce encapsulation and prevent direct access from outside

ali-safwat commented on lib/Feature/bestSellerFeature/data/repositories/best_seller_repo_impl.dart:
```diff
@@ -0,0 +1,17 @@
+
+@Injectable(as: BestSellerRepo)
+class BestSellerRepoImpl implements BestSellerRepo {
+  final BestSellerRemoteDataSource dataSource;
```
Make dataSource private inside this class to enforce encapsulation and prevent direct access from outside

ali-safwat commented on lib/Feature/bestSellerFeature/domain/useCases/best_seller_use_case.dart:
```diff
@@ -0,0 +1,14 @@
+
+@injectable
+class BestSellerUseCase{
+  final BestSellerRepo repo;
```
Make repo private inside this class to enforce encapsulation and prevent direct access from outside

ali-safwat reviewed (approved):
approved but need to solve conflict

YassenAhmed-gif committed:
Refactor: Update BestSeller feature and related components

This commit introduces several changes to the BestSeller feature, including model updates, UI enhancements, and test improvements.

Key changes include:
- Updated `BestSellerEntity` and `BestSellerDto` to include more fields like `slug`, `images`, `discount`, `rateAvg`, `rateCount`, `quantity`, `category`, `occasion`, `isSuperAdmin`, `createdAt`, `updatedAt`, `V`, and `sold`.
- Modified `BestSellerDtoMapper` to reflect the changes in `BestSellerDto` and `BestSellerEntity`.
- Updated data sources, repositories, and use cases to use private fields for dependencies.
- Enhanced `ProductCard` widget with `CachedNetworkImage` for better image loading and updated styling.
- Made `BestSellerScreen` responsive using `ResponsiveGridView` and updated product card data handling.
- Improved tests for `BestSellerEntity`, `BestSellerRemoteDataSourceImpl`, `BestSellerUseCase`, `BestSellerViewModel`, and mappers to cover new fields and scenarios.
- Added `preferredSize` to `CustomBackButton` in `custom_app_bar.dart`.
- Included `best_seller_dto_mapper.dart` in `best_seller_response_dto_mapper.dart`.
